### PR TITLE
add Python solution for Lifeguards (786)

### PR DESCRIPTION
**Pull request by jjliewie**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
finally someone else who's adding python support

**Comment by jjliewie** [comment]
> finally someone else who's adding python support

yeah idk why more people aren't using python in usaco. I'm not in plat yet so I wouldn't know about plat problems but you can totally solve silver & gold problems with python.

**Comment by SansPapyrus683** [comment]
not so sure about gold buddy

**Comment by jjliewie** [comment]
> not so sure about gold buddy

imma start adding python gold sols just to prove you wrong. usaco-guide finna have 20+ pull requests

**Comment by SansPapyrus683** [comment]
if you can solve [this](http://www.usaco.org/index.php?page=viewproblem2&cpid=970) in python i will literally be your slave for a week

**Comment by jjliewie** [comment]
> if you can solve [this](http://www.usaco.org/index.php?page=viewproblem2&cpid=970) in python i will literally be your slave for a week can't do anything overly weird tho

haven't tried this one yet but I'll solve it rn just for u <3 skipping school anyways why not

File: solutions/usaco-786.mdx [hunk 1]
```diff
@@ -240,4 +240,52 @@ public class lifeguards {
...
+if n == 1:
+	sys.stdout.write(str(0))
+	exit()
+
+sweep = []
+for _ in range(n):
+	a, b = map(int, t().split())
+	sweep.append([a, b])
+
+sweep.sort(key=lambda x:x[0])
```
Comment (ryanchou-dev) [inline solutions/usaco-786.mdx:265]:
```suggestion
sweep.sort(key=lambda x:x[0])
```
sorting in python automatically sorts by first element?

File: solutions/usaco-786.mdx [hunk 2]
```diff
@@ -240,4 +240,52 @@ public class lifeguards {
...
+	find_min[i] = sweep[i][1] - sweep[i][0]
+
+cnt = 1
+
+for start, end in sweep[1:]:
+	if temp > start:
+		cover += max(0, end-temp)
+		find_min[cnt-1] -= (temp - start) 
+		find_min[cnt] -= (min(temp, end) - start)
+	else: cover += end - start
```
Comment (ryanchou-dev) [inline solutions/usaco-786.mdx:281]: sep lines

**Review by ryanchou-dev — APPROVED** [review]
⭐

File: solutions/usaco-786.mdx (hunk 1 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:265]: yeah

File: solutions/usaco-786.mdx [hunk 3]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+<PySection>
+
+```python
+import sys
+
+sys.stdin = open("lifeguards.in", "r")
+sys.stdout = open("lifeguards.out", "w")
+
+t = sys.stdin.readline
+
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:253]: `with`

File: solutions/usaco-786.mdx [hunk 4]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+	
+temp = sweep[0][1]
+cover = sweep[0][1] - sweep[0][0]
+
+find_min = [0] * n
+
+for i in range(len(sweep)):
+	find_min[i] = sweep[i][1] - sweep[i][0]
+
+cnt = 1
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:275]: please use something other than `cnt`

File: solutions/usaco-786.mdx [hunk 5]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+
+for start, end in sweep[1:]:
+	if temp > start:
+		cover += max(0, end-temp)
+		find_min[cnt-1] -= (temp - start) 
+		find_min[cnt] -= (min(temp, end) - start)
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:281]: parens not necessary

File: solutions/usaco-786.mdx [hunk 6]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+find_min = [0] * n
+
+for i in range(len(sweep)):
+	find_min[i] = sweep[i][1] - sweep[i][0]
+
+cnt = 1
+
+for start, end in sweep[1:]:
+	if temp > start:
+		cover += max(0, end-temp)
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:279]: spaces for operators

File: solutions/usaco-786.mdx [hunk 7]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+		find_min[cnt-1] -= (temp - start) 
+		find_min[cnt] -= (min(temp, end) - start)
+	else: 
+		cover += end - start
+
+	temp = max(end, temp)
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:285]: better name

File: solutions/usaco-786.mdx [hunk 8]
```diff
@@ -240,4 +240,54 @@ public class lifeguards {
...
+
+t = sys.stdin.readline
+
+n = int(t())
+
+if n == 1:
+	sys.stdout.write(str(0))
+	exit()
+
+sweep = []
```
Comment (SansPapyrus683) [inline solutions/usaco-786.mdx:260]: comment for what this does

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-786.mdx (hunk 3 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:253]: okay

File: solutions/usaco-786.mdx (hunk 4 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:275]: alright

File: solutions/usaco-786.mdx (hunk 5 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:281]: okay ill delete

File: solutions/usaco-786.mdx (hunk 6 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:279]: yep alright

File: solutions/usaco-786.mdx (hunk 7 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:285]: right

File: solutions/usaco-786.mdx (hunk 4 above)
Comment (jjliewie) [reply solutions/usaco-786.mdx:270]: like what

Comment (SansPapyrus683) [reply solutions/usaco-786.mdx:270]: what's this counting?

Comment (jjliewie) [reply solutions/usaco-786.mdx:270]: the current index

Comment (SansPapyrus683) [reply solutions/usaco-786.mdx:270]: then `curr_index`

File: solutions/usaco-786.mdx [hunk 9]
```diff
@@ -240,4 +240,49 @@ public class lifeguards {
...
 
 </JavaSection>
+
+<PySection>
+
+```python
+sweep = []
+	
+with open('lifeguards.in') as r:
+	t = r.readline
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-786.mdx:250]:
```suggestion
```

File: solutions/usaco-786.mdx [hunk 10]
```diff
@@ -240,4 +240,49 @@ public class lifeguards {
...
+	n = int(t())
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-786.mdx:251]:
```suggestion
	n = int(r.readline())
```

File: solutions/usaco-786.mdx [hunk 11]
```diff
@@ -240,4 +240,49 @@ public class lifeguards {
...
+	for _ in range(n):
+		a, b = map(int, t().split())
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-786.mdx:253]:
```suggestion
		a, b = map(int, r.readline().split())
```

File: solutions/usaco-786.mdx [hunk 12]
```diff
@@ -240,4 +240,49 @@ public class lifeguards {
 ```
 
 </JavaSection>
+
+<PySection>
+
+```python
+sweep = []
+	
+with open('lifeguards.in') as r:
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-786.mdx:249]: change `r` to `fin` or smth

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-786.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/usaco-786.mdx:250]: agree with this change

File: solutions/usaco-786.mdx (hunk 12 above)
Comment (SansPapyrus683) [reply solutions/usaco-786.mdx:249]: (optional)

Comment (jjliewie) [reply solutions/usaco-786.mdx:249]: okay

File: solutions/usaco-786.mdx [hunk 13]
```diff
@@ -240,4 +240,48 @@ public class lifeguards {
...
+		find_min[cnt] -= min(cur, end) - start
+	else: 
+		cover += end - start
+
+	cur = max(end, cur)
+	cnt += 1
+
+ans = cover - max(min(find_min), 0)
+
+print(str(ans))
```
Comment (chezbgone) [inline solutions/usaco-786.mdx:284]: The `print` function automatically converts the argument a string, so `print(ans)` is fine.

File: solutions/usaco-786.mdx [hunk 14]
```diff
@@ -240,4 +240,48 @@ public class lifeguards {
...
+
+sweep.sort()
+	
+cur = sweep[0][1]
+cover = sweep[0][1] - sweep[0][0]
+
+find_min = [0] * n
+
+for i in range(len(sweep)):
+	find_min[i] = sweep[i][1] - sweep[i][0]
```
Comment (chezbgone) [inline solutions/usaco-786.mdx:267]: More idiomatic to write
`find_min = [r - l for (l, r) in sweep]`

**Review by chezbgone — CHANGES_REQUESTED** [review]

**Comment by chezbgone** [comment]
There are also a few other requested changes from other reviewers. You can see all of the comments in the `Files changed` tab of this pull request.

**Pull request closed without merging**